overlord/snapstate/refreshhints.go

```diff
@@ -121,3 +136,52 @@ func (r *refreshHints) Ensure() error {
 	}
 	return r.refresh()
 }
+
+func refreshHintsFromCandidates(st *state.State, updates []*snap.Info, ignoreValidationByInstanceName map[string]bool, deviceCtx DeviceContext) ([]*refreshCandidate, error) {
+	if ValidateRefreshes != nil && len(updates) != 0 {
+		userID := 0
+		var err error
+		updates, err = ValidateRefreshes(st, updates, ignoreValidationByInstanceName, userID, deviceCtx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	hints := []*refreshCandidate{}
+	for _, update := range updates {
+		var snapst SnapState
+		if err := Get(st, update.InstanceName(), &snapst); err != nil {
+			return nil, err
+		}
+
+		flags := snapst.Flags
+		flags.IsAutoRefresh = true
+		flags, err := earlyChecks(st, &snapst, update, flags)
+		if err != nil {
+			logger.Noticef("cannot update %q: %v", update.InstanceName(), err)
+			continue
+		}
+
+		if err := earlyEpochCheck(update, &snapst); err != nil {
+			logger.Noticef("cannot update %q: %v", update.InstanceName(), err)
+			continue
+		}
+		snapsup := &refreshCandidate{
+			SnapSetup{
```

Comment:
could you order this in the same order they are set in doUpdate,  with comments for the one we don't set, I think some are set then in MakeSnapSetup  which is fine but I don't know what is the plan for UserID ?

Reply:
Done. UserID will be 0 for auto-refresh (and it is not relevant at this point). I've modified this slightly to also set cohort and channel so they are available for snapctl side without needing to consult snapstate.